velox/functions/lib/Re2Functions.cpp

```diff
@@ -1119,120 +1246,185 @@ class PatternStringIterator {
   std::string_view pattern_;
   const std::optional<char> escapeChar_;
 
+  // Index of current char(including the escape char if there is).
   size_t currentStart_{0};
+  // Index of next char.
   size_t nextStart_{0};
   CharKind charKind_{CharKind::kNormal};
-  bool isPreviousWildcard_{false};
 };
 
 PatternMetadata determinePatternKind(
     std::string_view pattern,
     std::optional<char> escapeChar) {
-  const size_t patternLength = pattern.size();
-
-  // Index of the first % or _ character(not escaped).
-  int32_t wildcardStart = -1;
-  // Index of the first character that is not % and not _.
-  int32_t fixedPatternStart = -1;
-  // Index of the last character in the fixed pattern, used to retrieve the
-  // fixed string for patterns of type kSubstring.
-  int32_t fixedPatternEnd = -1;
-  // Count of wildcard character sequences in pattern.
-  size_t numWildcardSequences = 0;
-  // Total number of % characters.
-  size_t anyCharacterWildcardCount = 0;
-  // Total number of _ characters.
-  size_t singleCharacterWildcardCount = 0;
-
+  std::vector<SubPatternKind> subPatternKinds;
+  std::vector<std::pair<size_t, size_t>> subPatternRanges;
   PatternStringIterator iterator{pattern, escapeChar};
 
   // Iterate through the pattern string to collect the stats for the simple
   // patterns that we can optimize.
+  std::ostringstream os;
+  SubPatternKind previousKind;
+  size_t currentSubPatternStart = 0;
+  size_t cursor = 0;
   while (iterator.next()) {
```

Comment:
Does this logic handle empty pattern properly?

Reply:
Added check to categorize empty pattern as kFixed("").